### PR TITLE
Build: Packaging for driver-embedded projects

### PR DESCRIPTION
**Pull request by rozza**
* Refactored how plugins are applied across the projects.
  As the android plugin can't be used if the java plugin
  is installed.
* Added driver-embedded-android project to handle building aar.
* Updated gradle and plugin versions

JAVA-2929 JAVA-2930

### CODE REVIEW

**Comment by rozza** [comment]
https://evergreen.mongodb.com/version/5bb49baf0305b91647c30f95

File: gradle/publish.gradle [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+                    password project.hasProperty('nexusPassword') ? project.getProperty('nexusPassword') : ''
+                }
+            }
+        }
+    }
+    configureSigning = { project ->
+        { ->
+            def isReleaseVersion = !project.version.endsWith('SNAPSHOT')
+            signing {
+                required { isReleaseVersion && gradle.taskGraph.hasTask('uploadArchives') }
```
Comment (jyemin) [inline gradle/publish.gradle]: I don't think I understand this.  This seems to indicate that we only sign release versions, but we do also sign snapshots.

File: bson/build.gradle [hunk 2]
```diff
@@ -14,12 +14,13 @@
...
 
-apply plugin: 'osgi'
-apply plugin: 'java'
-apply plugin: 'org.kordamp.gradle.clirr'
-
-def configDir = new File(rootDir, 'config')
 archivesBaseName = 'bson'
+description = 'The BSON library'
+
+ext {
```
[the remark this answers is not in the document]
Comment (rozza) [reply bson/build.gradle]: This is the documented way of setting ext variables. 

The only reason we need them per project is to be able to configure the pom as there are no post configuration hooks like we previously had with the older plugin and `modifyPom`

File: driver-embedded-android/build.gradle [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+dependencies {
+    api project(':driver-sync')
+    implementation "org.mongodb:embedded-capi-android:[4.0,4.1)@aar"
+}
+
+android {
+    compileSdkVersion 28
+
+    defaultConfig {
+        minSdkVersion 24
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-embedded-android/build.gradle]: We follow how the server is packaged: See https://github.com/mongodb/mongo/blob/747a005fb46d3432d576fe56051b0f476e2d2c96/src/mongo/embedded/mongo_embedded/java/aar/build.gradle

File: driver-embedded-android/build.gradle [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        targetSdkVersion 28
+    }
+
+    buildTypes {
+        release {
+            minifyEnabled false
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-embedded-android/build.gradle]: Follows the server convention - ensures the aar doesn't go through a proGuard minification step.

File: driver-embedded-android/build.gradle [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    sourceSets {
+        main {
+            manifest.srcFile 'AndroidManifest.xml'
+            java.srcDir "$rootDir/driver-embedded/src/main"
+        }
+    }
+
+    packagingOptions {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-embedded-android/build.gradle]: This is no longer needed - building the aar previously failed but looks like further refactoring has negated the need for this block.

File: build.gradle [hunk 6]
```diff
@@ -14,133 +14,101 @@
...
     dependencies {
         compileOnly 'com.google.code.findbugs:jsr305:1.3.9'
         compile 'org.slf4j:slf4j-api:1.7.6', optional
-
         testCompile 'com.google.code.findbugs:jsr305:1.3.9'
     }
 
     /* Compiling */
     tasks.withType(AbstractCompile) {
+        onlyIf { JavaVersion.current().isJava7Compatible() }
```
[the remark this answers is not in the document]
Comment (rozza) [reply build.gradle]: Ack

File: driver-embedded-android/build.gradle [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    classpath += configurations.compile
+    destinationDir = new File(project.buildDir, 'docs')
+    rootProject.tasks.withType(Javadoc).each { javadocTask ->
+        source += javadocTask.source
+        classpath += javadocTask.classpath
+        excludes += javadocTask.excludes
+        includes += javadocTask.includes
+        dependsOn += javadocTask.dependsOn
+    }
+    exclude "**/com/mongodb/**/benchmark/**"
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-embedded-android/build.gradle]: Yes.

File: gradle/publish.gradle (hunk 1 above)
Comment (rozza) [reply gradle/publish.gradle]: For some reason I thought we didn't sign the snapshots - I'll remove this. Looks like it only signs the main jar - so will need to review and ensure the javadoc & poms also get signed.

File: bson/build.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply bson/build.gradle]: Acknowledged.  I had posted this question before our walkthrough and neglected to delete it.

File: driver-embedded-android/build.gradle (hunk 4 above)
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-embedded-android/build.gradle]: Acknowledged

**Comment by jyemin** [comment]
Very nice. LGTM, pending Stitch sign-off.

File: driver-embedded-android/build.gradle (hunk 3 above)
[the remark this answers is not in the document]
Comment (edaniels) [reply driver-embedded-android/build.gradle]: If necessary Stitch can bump to 24. Could one of you confirm that server can't go to 21. Pretty sure you can't

[the remark this answers is not in the document]
Comment (henrikedin) [reply driver-embedded-android/build.gradle]: I think we can go down to 21 now when the library collapse has landed. I'll let you know.

**Comment by edaniels** [comment]
@rozza This all works for Stitch! We just want to go down to API Level 21 if @henrikedin says we can and his team publishes the AAR

**Comment by rozza** [comment]
@edaniels I've updated the min version and a new snapshot is now in sonatype.

**Comment by edaniels** [comment]
Perfect. Thanks Ross!

**Pull request closed without merging**